File: AI/middleware/loop_detection.py
"""
第六层: 检测循环行为 (LoopDetectionMiddleware)

问题场景:
- 智能体反复执行相同行动
- 智能体陷入死循环

解决方案:
- 跟踪行动历史,检测重复模式
"""

from typing import Any, Dict, List
from collections import Counter

from .base import BaseMiddleware, MiddlewareContext, MiddlewareResult


class LoopDetectionMiddleware(BaseMiddleware):
    """检测循环行为中间件"""

    def __init__(self,
                 warn_threshold: int = 3,
                 hard_limit: int = 5,
                 window_size: int = 20):
        super().__init__("LoopDetectionMiddleware")
        self.warn_threshold = warn_threshold
        self.hard_limit = hard_limit
        self.window_size = window_size
        self.action_history: Dict[str, List[str]] = {}

    def process(self, context: MiddlewareContext, agent: Any) -> MiddlewareResult:
        if not self.enabled:
            return MiddlewareResult()

        agent_id = context.agent_id
        action = context.action or context.goal or ""

        if not action:
            return MiddlewareResult()

        # Record action
        if agent_id not in self.action_history:
            self.action_history[agent_id] = []

        self.action_history[agent_id].append(action)

        # Keep window size
        if len(self.action_history[agent_id]) > self.window_size:
            self.action_history[agent_id] = self.action_history[agent_id][-self.window_size:]

        # Check for loops
        recent = self.action_history[agent_id]
        if len(recent) >= self.warn_threshold:
            counter = Counter(recent[-self.warn_threshold:])
            most_common = counter.most_common(1)[0]
            count = most_common[1]

            if count >= self.hard_limit:
                return MiddlewareResult(
                    should_continue=False,
                    modified_goal="explore",
                    message=f"[{self.name}] 检测到死循环,强制变更目标: {most_common[0]} x{count}"
                )

            if count >= self.warn_threshold:
                return MiddlewareResult(
                    should_continue=True,
                    modified_goal="explore",
                    message=f"[{self.name}] 警告: 可能存在循环行为 - {most_common[0]} x{count}"
                )

        return MiddlewareResult()

File: src/AI/middleware/loop_detection.ts
/**
 * 循环检测中间件
 * 用于检测智能体是否在重复执行相同的动作，防止无限循环
 */
import { BaseMiddleware, MiddlewareContext, MiddlewareResult } from './base';

export class LoopDetectionMiddleware extends BaseMiddleware {
  // 最近执行的动作历史（按 agent_id 分组）
  private recentActions: Map<string, string[]>;
  // 最大历史记录长度
  private maxHistory: number;
  // 连续重复次数阈值
  private consecutiveThreshold: number;

  /**
   * 构造函数
   * @param maxHistory 最大历史记录长度，默认为10
   * @param consecutiveThreshold 连续重复阈值，默认为3
   */
  constructor(maxHistory: number = 10, consecutiveThreshold: number = 3) {
    super("LoopDetectionMiddleware");
    this.recentActions = new Map();
    this.maxHistory = maxHistory;
    this.consecutiveThreshold = consecutiveThreshold;
  }

  /**
   * 处理方法
   * 检测是否存在循环动作
   * @param context 中间件上下文
   * @param agent 智能体实例
   * @returns 中间件结果
   */
  process(context: MiddlewareContext, agent: any): MiddlewareResult {
    // 检查是否有动作
    if (context.action) {
      const agentId = context.agent_id || 'unknown';
      
      // 获取该智能体的动作历史
      if (!this.recentActions.has(agentId)) {
        this.recentActions.set(agentId, []);
      }
      const actions = this.recentActions.get(agentId)!;
      
      // 检查连续重复次数
      let consecutiveCount = 0;
      for (let i = actions.length - 1; i >= 0; i--) {
        if (actions[i] === context.action) {
          consecutiveCount++;
        } else {
          break;
        }
      }
      
      // 如果连续重复次数超过阈值，检测到循环
      if (consecutiveCount >= this.consecutiveThreshold) {
        console.warn(`[LoopDetection] 智能体 ${agentId} 连续 ${consecutiveCount} 次执行相同动作`);
        return {
          should_continue: false,
          message: "检测到循环动作，已停止执行",
          metadata: {
            loop_detected: true,
            action: context.action,
            consecutive_count: consecutiveCount
          }
        };
      }

      // 将动作添加到历史记录
      actions.push(context.action);
      // 如果历史记录超过最大长度，移除最早的记录
      if (actions.length > this.maxHistory) {
        actions.shift();
      }
    }

    // 没有检测到循环，继续执行
    return {
      should_continue: true,
      metadata: {}
    };
  }
}
